bobtimus/src/bin/bobtimus.rs

```diff
@@ -0,0 +1,204 @@
+use anyhow::Result;
+use bobtimus::{cli::StartCommand, routes, Bobtimus};
+use elements_fun::secp256k1::rand::{rngs::StdRng, thread_rng, SeedableRng};
+use elements_harness::Client;
+use structopt::StructOpt;
+use warp::Filter;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    env_logger::init();
+
+    let StartCommand {
+        elementsd_url,
+        api_port,
+        usdt_asset_id,
+    } = StartCommand::from_args();
+
+    let elementsd = Client::new(elementsd_url.into_string())?;
+    let btc_asset_id = elementsd.get_bitcoin_asset_id().await?;
+
+    let rate_service = kraken::RateService::new().await?;
+
+    let bobtimus = Bobtimus {
+        rng: StdRng::from_rng(&mut thread_rng()).unwrap(),
+        rate_service,
+        elementsd,
+        btc_asset_id,
+        usdt_asset_id,
+    };
+
+    let bobtimus_filter = warp::any().map({
+        let bobtimus = bobtimus.clone();
+        move || bobtimus.clone()
+    });
+
+    let subscription = bobtimus.rate_service.subscribe();
+    let latest_rate = warp::path!("rate" / "lbtc-lusdt")
+        .and(warp::get())
+        .map(move || routes::latest_rate(subscription.clone()));
+
+    let create_swap = warp::post()
+        .and(warp::path!("swap" / "lbtc-lusdt"))
+        .and(warp::path::end())
+        .and(bobtimus_filter.clone())
+        .and(warp::body::json())
+        .and_then(routes::create_swap);
+
+    warp::serve(latest_rate.or(create_swap))
+        .run(([127, 0, 0, 1], api_port))
+        .await;
+
+    Ok(())
+}
+
+mod kraken {
+    use anyhow::{anyhow, bail, Result};
+    use async_trait::async_trait;
+    use bobtimus::{LatestRate, LiquidUsdt, Rate};
+    use futures::{SinkExt, Stream, StreamExt};
+    use reqwest::Url;
+    use serde::{Deserialize, Serialize};
+    use serde_json::Value;
+    use std::convert::TryFrom;
+    use tokio::sync::watch;
+    use tokio_tungstenite::tungstenite::Message;
+    use watch::Receiver;
+
+    const KRAKEN_WS_URL: &str = "wss://ws.kraken.com";
+    const SUBSCRIBE_XBT_USDT_TICKER_PAYLOAD: &str = r#"
+    { "event": "subscribe",
+      "pair": [ "XBT/USD" ],
+      "subscription": {
+        "name": "ticker"
+      }
+    }"#;
+
+    #[derive(Clone)]
+    pub struct RateService(Receiver<Rate>);
+
+    #[async_trait]
+    impl LatestRate for RateService {
+        async fn latest_rate(&mut self) -> anyhow::Result<Rate> {
```

Comment:
So this implementation will currently block on the next rate before it can proceed whereas ideally, we want it to use the last one we sent to the user.

Instead of having this functionality here, you could just make another struct that - given a stream of rates - always stores the latest value in a local field and makes it available through a getter. (For that you will need spawn a task with a loop into the runtime)

You might want to use a `RwLock` for that instead of a `Mutex` (so that different clients don't block each other).

Reply:
I've solved it (I think) by implementing `Future` on `RateService` and always looking for the last value in the stream, updating an internal state and only if the future is pending actually returning the stored state. Take a look: https://github.com/comit-network/droplet/pull/28/commits/d1f13225d7ee40209c05cb649a5a191e07f0c48b.

Reply:
Interesting solution! :)